src/x509_str.c

```diff
@@ -300,17 +286,170 @@ int wolfSSL_X509_verify_cert(WOLFSSL_X509_STORE_CTX* ctx)
+                break;
+            }
+
+            /* We found our issuer in the non-trusted cert list, add it
+             * to the CM and verify the current cert against it */
+            ret = X509StoreAddCa(ctx->store, issuer,
+                                            WOLFSSL_INTER_CA);
+            if (ret != WOLFSSL_SUCCESS) {
+                goto exit;
+            }
+
+            added = 1;
+
+            ret = X509StoreVerifyCert(ctx);
+            if (ret != WOLFSSL_SUCCESS) {
+                goto exit;
+            }
+
+            /* Add it to the current chain and look at the issuer cert next */
+            wolfSSL_sk_X509_push(ctx->chain, ctx->current_cert);
+            ctx->current_cert = issuer;
+        }
+        else if (ret == WC_NO_ERR_TRACE(WOLFSSL_FAILURE)) {
+            /* Could not find in untrusted list, only place left is
+             * a trusted CA in the CM */
+            ret = X509StoreVerifyCert(ctx);
+            if (ret != WOLFSSL_SUCCESS) {
+                if (((ctx->flags & X509_V_FLAG_PARTIAL_CHAIN) ||
+                     (ctx->store->param->flags & X509_V_FLAG_PARTIAL_CHAIN)) &&
+                    (added == 1)) {
+                    wolfSSL_sk_X509_push(ctx->chain, ctx->current_cert);
+                    ret = WOLFSSL_SUCCESS;
+                }
+                goto exit;
+            }
+
+            /* Cert verified, finish building the chain */
+            wolfSSL_sk_X509_push(ctx->chain, ctx->current_cert);
+    #ifdef WOLFSSL_SIGNER_DER_CERT
+            x509GetIssuerFromCM(&issuer, ctx->store->cm, ctx->current_cert);
+            if (issuer != NULL && ctx->store->owned != NULL) {
+                wolfSSL_sk_X509_push(ctx->store->owned, issuer);
+            }
+    #else
+            if (ctx->setTrustedSk == NULL) {
+                X509StoreGetIssuerEx(&issuer,
+                    ctx->store->trusted, ctx->current_cert);
+            }
+            else {
+                X509StoreGetIssuerEx(&issuer,
+                    ctx->setTrustedSk, ctx->current_cert);
+            }
+    #endif
+            if (issuer != NULL) {
+                wolfSSL_sk_X509_push(ctx->chain, issuer);
+            }
+
+            done = 1;
+        }
+        else {
+            goto exit;
+        }
+
+        depth--;
+    }
+
+exit:
+    /* Remove additional intermediates from init from the store */
+    if (ctx != NULL && numInterAdd > 0) {
+        for (i = 0; i < numInterAdd; i++) {
+            wolfSSL_sk_X509_pop(ctx->store->certs);
+        }
+    }
+    /* Remove intermediates that were added to CM */
+    if (ctx != NULL) {
+        if (ctx->store != NULL) {
+            if (added == 1) {
+                wolfSSL_CertManagerUnloadTempIntermediateCerts(ctx->store->cm);
+            }
+        }
+        if (orig != NULL) {
+            ctx->current_cert = orig;
+        }
+    }
+
+    return ret == WOLFSSL_SUCCESS ? WOLFSSL_SUCCESS : WOLFSSL_FAILURE;
```

Comment:
Can you confirm you only want to return success or failure and not any other possible return codes? Old logic was >= 0 success else failure.
@douzzer does this need to be `WC_NO_ERR_TRACE(WOLFSSL_FAILURE)`, etc?

Reply:
Talked with douzzer, he said this should be OK.

Reply:
@ColtonWilley the first half of the question still stands -- do we really want to be masking the true return code here?

Reply:
Yes actually I believe we do. I believe openssl only returns 0 or 1, and uses `X509_STORE_CTX_get_error` for error reporting.

Reply:
OpenSSL returns `-1` on an error, `0` on verification failure, and `1` on verification success. This rework conforms to that API.